Cannot connect to new MongoDB source or sync existing databases

### Describe the bug

I've recently upgraded Metabase from v0.47.9 to v0.48.3 and I've noticed that we even though we could ask for a sync to the database schema through `Admin -> Databases -> Sync database schema now`, there was no change in the `Table Metadata`.

In the application logs, I've noticed that something was off:
```
2024-01-22 18:37:05,509 ERROR driver.util :: Failed to connect to Database
java.lang.Exception: Failed to connect to Database
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__57039.invoke(util.clj:147)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2024-01-22 18:37:05,511 WARN middleware.log :: POST /api/database/161/sync_schema 422 46.9 ms (3 chamadas ao banco de dados) 
"Failed to connect to Database"
```

Investigating further, I noticed that our Postgres databases could sync without errors, but I couldn't sync any of our local Mongo databases(local cluster). I've assumed that it could be either one of three problems:
- Network issues
- Mongo cluster issues
- Metabase issues

Updating data from Mongo still worked through `Admin -> Databases -> Re-scan field values now`, so this indicated that it did not appear to be a problem connecting to the Mongo cluster, and also no network issue between them.

In order to test if a specific version induced this problem, I've created the Metabase instance in my local dev environment via docker compose , and tried to connect to a MongoDB (also a local docker container).
The following versions were tested:
| Version | Connected to MongoDB |
|--------| :---: |
| v.0.47.9 | Yes |
| v0.47.10 | Yes |
| v0.48.0 | No |
| v0.48.1 | No |
| v0.48.2 | No |
 v0.48.3 | No |

### To Reproduce

1. Deploy an instance of `metabase/metabase:v0.48.3` image via docker
2. Try to add a new MongoDB source via `Admin -> Databases`


### Expected behavior

Should connect to MongoDB

### Logs

Error log received after creating a new connection
```
2024-01-22 20:49:58,938 ERROR driver.util :: Failed to connect to Database
java.lang.Exception: Failed to connect to Database
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__57039.invoke(util.clj:147)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2024-01-22 20:49:58,976 ERROR driver.util :: Failed to connect to Database
java.lang.Exception: Failed to connect to Database
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__57039.invoke(util.clj:147)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2024-01-22 20:49:58,978 ERROR api.database :: Não é possível conectar ao banco de dados
clojure.lang.ExceptionInfo: Failed to connect to Database {:message "Failed to connect to Database"}
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:164)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:135)
	at clojure.lang.RestFn.invoke(RestFn.java:442)
	at metabase.api.database$test_database_connection.invokeStatic(database.clj:694)
	at metabase.api.database$test_database_connection.doInvoke(database.clj:684)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at metabase.api.database$fn__99799$_AMPERSAND_f__99800.invoke(database.clj:748)
	at metabase.api.database$fn__99799$fn__99807.invoke(database.clj:728)
	at metabase.api.database$fn__99820.invokeStatic(database.clj:770)
	at metabase.api.database$fn__99820.invoke(database.clj:752)
	at compojure.core$wrap_response$fn__43924.invoke(core.clj:160)
	at compojure.core$wrap_route_middleware$fn__43908.invoke(core.clj:132)
	at compojure.core$wrap_route_info$fn__43913.invoke(core.clj:139)
	at compojure.core$wrap_route_matches$fn__43917.invoke(core.clj:151)
	at clojure.lang.Var.invoke(Var.java:393)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__43917.invoke(core.clj:152)
	at clojure.lang.Var.invoke(Var.java:393)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936.invoke(core.clj:200)
	at metabase.server.middleware.auth$enforce_authentication$fn__95229.invoke(auth.clj:17)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936.invoke(core.clj:200)
	at compojure.core$make_context$handler__43964.invoke(core.clj:290)
	at compojure.core$make_context$fn__43968.invoke(core.clj:300)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$make_context$fn__43968.invoke(core.clj:301)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at metabase.api.routes$fn__102992$fn__102995.invoke(routes.clj:66)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936.invoke(core.clj:200)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.server.routes$fn__103155$fn__103156.doInvoke(routes.clj:72)
	at clojure.lang.RestFn.invoke(RestFn.java:436)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936.invoke(core.clj:200)
	at compojure.core$make_context$handler__43964.invoke(core.clj:290)
	at compojure.core$make_context$fn__43968.invoke(core.clj:300)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__43917.invoke(core.clj:153)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__43917.invoke(core.clj:153)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__43917.invoke(core.clj:153)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936$f__43937$respond_SINGLEQUOTE___43938.invoke(core.clj:197)
	at metabase.server.routes$fn__103140$fn__103142.invoke(routes.clj:49)
	at compojure.core$routes$fn__43936$f__43937.invoke(core.clj:198)
	at compojure.core$routes$fn__43936.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__100042.invoke(exceptions.clj:108)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__100039.invoke(exceptions.clj:96)
	at metabase.server.middleware.log$log_api_call$fn__103421$fn__103422$fn__103423.invoke(log.clj:216)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:18)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__103421$fn__103422.invoke(log.clj:208)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__103421.invoke(log.clj:207)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__105454.invoke(browser_cookie.clj:40)
	at metabase.server.middleware.security$add_security_headers$fn__85933.invoke(security.clj:180)
	at metabase.server.middleware.json$wrap_json_body$fn__45297.invoke(json.clj:67)
	at metabase.server.middleware.offset_paging$handle_paging$fn__85957.invoke(offset_paging.clj:45)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__45315.invoke(json.clj:103)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__105721.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__105740.invoke(params.clj:77)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__66045.invoke(misc.clj:61)
	at metabase.server.middleware.session$reset_session_timeout$fn__72112.invoke(session.clj:488)
	at metabase.server.middleware.session$bind_current_user$fn__72079$fn__72080.invoke(session.clj:383)
	at metabase.server.middleware.session$do_with_current_user.invokeStatic(session.clj:362)
	at metabase.server.middleware.session$do_with_current_user.invoke(session.clj:346)
	at metabase.server.middleware.session$bind_current_user$fn__72079.invoke(session.clj:382)
	at metabase.server.middleware.session$wrap_current_user_info$fn__72062.invoke(session.clj:321)
	at metabase.server.middleware.session$wrap_session_id$fn__72045.invoke(session.clj:253)
	at metabase.server.middleware.auth$wrap_api_key$fn__95237.invoke(auth.clj:30)
	at ring.middleware.cookies$wrap_cookies$fn__105641.invoke(cookies.clj:216)
	at metabase.server.middleware.misc$add_content_type$fn__66027.invoke(misc.clj:29)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__66053.invoke(misc.clj:78)
	at ring.middleware.gzip$wrap_gzip$fn__105683.invoke(gzip.clj:86)
	at metabase.server.middleware.misc$bind_request$fn__66056.invoke(misc.clj:95)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__105470.invoke(ssl.clj:51)
	at metabase.server$async_proxy_handler$fn__66230.invoke(server.clj:77)
	at metabase.server.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1598)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:753)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:501)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:287)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:199)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.Exception: Failed to connect to Database
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__57039.invoke(util.clj:147)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
2024-01-22 20:49:58,987 DEBUG middleware.log :: POST /api/database 400 97.8 ms (0 chamadas ao banco de dados) 
{:message "Failed to connect to Database"}
```

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.21+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.21",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.21+9",
    "os.name": "Linux",
    "os.version": "4.9.0-9-amd64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mongo",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "15.5 (Debian 15.5-1.pgdg110+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.6.0"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-01-11",
      "tag": "v0.48.3",
      "hash": "80d8323"
    },
    "settings": {
      "report-timezone": "America/Sao_Paulo"
    }
  }
}
```


### Severity

Blocks some use cases and updating our applications

### Additional context

Our production Metabase instance is deployed via Docker Compose:
```yaml
services:
  metabase:
    image: metabase/metabase:v0.48.3
    hostname: metabase
    volumes:
      - /dev/urandom:/dev/random:ro
    ports:
      - 3000:3000
    environment:
      MB_DB_TYPE: postgres
      MB_DB_DBNAME: metabase
      MB_DB_PORT: 5432
      MB_DB_USER_FILE: /run/secrets/db_user
      MB_DB_PASS_FILE: /run/secrets/db_password
      MB_DB_HOST: postgres
    networks:
      - metanet1
    secrets:
      - db_password
      - db_user
    depends_on:
      postgres:
        condition: service_healthy
    healthcheck:
      test: curl --fail -I http://localhost:3000/api/health || exit 1
      interval: 15s
      timeout: 5s
      retries: 5
    restart: unless-stopped

  postgres:
    image: postgres:15.5-bullseye
    hostname: postgres
    environment:
      POSTGRES_USER_FILE: /run/secrets/db_user
      POSTGRES_DB: metabase
      POSTGRES_PASSWORD_FILE: /run/secrets/db_password
    volumes:
      - pg_metabase:/var/lib/postgresql/data
    ports:
      - 5442:5432
    healthcheck:
      test: ["CMD-SHELL", "pg_isready"]
      interval: 10s
      timeout: 5s
      retries: 5

    networks:
      - metanet1
    secrets:
      - db_password
      - db_user
    restart: unless-stopped

volumes:
  pg_metabase:

networks:
  metanet1:
    driver: bridge
secrets:
  db_password:
    file: db_password.txt
  db_user:
    file: db_user.txt
```

--- a/.github/workflows/drivers.yml
+++ b/.github/workflows/drivers.yml
@@ -281,7 +281,7 @@ jobs:
       uses: ./.github/actions/test-driver
       with:
         junit-name: 'be-tests-mongo-4-4-ee'
-        test-args: ":exclude-tags '[:mb/once]'"
+        test-args: ":exclude-tags '[:mb/once :mongo-sharded-cluster-tests]'"
     - name: Upload Test Results
       uses: ./.github/actions/upload-test-results
       if: always()
@@ -327,7 +327,7 @@ jobs:
         uses: ./.github/actions/test-driver
         with:
           junit-name: 'be-tests-mongo-4-4-ee'
-          test-args: ":exclude-tags '[:mb/once]'"
+          test-args: ":exclude-tags '[:mb/once :mongo-sharded-cluster-tests]'"
       - name: Upload Test Results
         uses: ./.github/actions/upload-test-results
         if: always()
@@ -361,7 +361,7 @@ jobs:
       uses: ./.github/actions/test-driver
       with:
         junit-name: 'be-tests-mongo-5-0-ee'
-        test-args: ":exclude-tags '[:mb/once]'"
+        test-args: ":exclude-tags '[:mb/once :mongo-sharded-cluster-tests]'"
     - name: Upload Test Results
       uses: ./.github/actions/upload-test-results
       if: always()
@@ -407,7 +407,7 @@ jobs:
       uses: ./.github/actions/test-driver
       with:
         junit-name: 'be-tests-mongo-5-0-ee'
-        test-args: ":exclude-tags '[:mb/once]'"
+        test-args: ":exclude-tags '[:mb/once :mongo-sharded-cluster-tests]'"
     - name: Upload Test Results
       uses: ./.github/actions/upload-test-results
       if: always()
@@ -444,7 +444,39 @@ jobs:
       uses: ./.github/actions/test-driver
       with:
         junit-name: 'be-tests-mongo-latest-ee'
-        test-args: ":exclude-tags '[:mb/once]'"
+        test-args: ":exclude-tags '[:mb/once :mongo-sharded-cluster-tests]'"
+    - name: Upload Test Results
+      uses: ./.github/actions/upload-test-results
+      if: always()
+      with:
+        input-path: ./target/junit/
+        output-name: ${{ github.job }}
+        bucket: ${{ vars.AWS_S3_TEST_RESULTS_BUCKET }}
+        aws-access-key-id: ${{ secrets.AWS_TEST_RESULTS_ACCESS_KEY_ID }}
+        aws-secret-access-key: ${{ secrets.AWS_TEST_RESULTS_SECRET_ACCESS_KEY }}
+        aws-region: ${{ vars.AWS_REGION }}
+        trunk-api-token: ${{ secrets.TRUNK_API_TOKEN }}
+
+  be-tests-mongo-sharded-cluster-ee:
+    needs: files-changed
+    if: github.event.pull_request.draft == false && needs.files-changed.outputs.backend_all == 'true'
+    runs-on: ubuntu-22.04
+    timeout-minutes: 90
+    env:
+      CI: 'true'
+      DRIVERS: mongo
+    services:
+      mongodb:
+        image: metabase/mongo-sharded-cluster:6
+        ports:
+          - "27017:17017"
+    steps:
+    - uses: actions/checkout@v4
+    - name: Test MongoDB driver (Sharded cluster)
+      uses: ./.github/actions/test-driver
+      with:
+        junit-name: 'be-tests-mongo-sharded-cluster-ee'
+        test-args: ':only "[metabase.driver.mongo.sharded-cluster-test]"'
     - name: Upload Test Results
       uses: ./.github/actions/upload-test-results
       if: always()

--- a/modules/drivers/mongo/src/metabase/driver/mongo.clj
+++ b/modules/drivers/mongo/src/metabase/driver/mongo.clj
@@ -49,7 +49,8 @@
   [_ db-details]
   (mongo.connection/with-mongo-client [^MongoClient c db-details]
     (let [db-names (mongo.util/list-database-names c)
-          db (mongo.util/database c (mongo.db/db-name db-details))
+          db-name (mongo.db/db-name db-details)
+          db (mongo.util/database c db-name)
           db-stats (mongo.util/run-command db {:dbStats 1} :keywordize true)]
       (and
        ;; 1. check db.dbStats command completes successfully
@@ -57,7 +58,7 @@
           1.0)
        ;; 2. check the database is actually on the server
        ;; (this is required because (1) is true even if the database doesn't exist)
-       (boolean (some #(= % (:db db-stats)) db-names))))))
+       (boolean (some #(= % db-name) db-names))))))
 
 (defmethod driver/humanize-connection-error-message
   :mongo

--- a/modules/drivers/mongo/test/metabase/driver/mongo/sharded_cluster_test.clj
+++ b/modules/drivers/mongo/test/metabase/driver/mongo/sharded_cluster_test.clj
@@ -1,0 +1,13 @@
+(ns ^:mongo-sharded-cluster-tests metabase.driver.mongo.sharded-cluster-test
+  (:require
+   [clojure.test :refer :all]
+   [metabase.driver :as driver]
+   [metabase.test :as mt]))
+
+(set! *warn-on-reflection* true)
+
+(deftest ^:synchronized can-connect-test
+  (mt/test-driver
+   :mongo
+   (testing "Mongo driver can connect to a sharded cluster"
+     (is (true? (driver/can-connect? :mongo (mt/db)))))))

